app/upgrades/v17/upgrades.go

```diff
@@ -1,105 +1,66 @@
 package v17
 
 import (
+	"errors"
 	"fmt"
 	"time"
 
+	errorsmod "cosmossdk.io/errors"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/types/module"
 	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
 
 	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
 
 	cltypes "github.com/osmosis-labs/osmosis/v17/x/concentrated-liquidity/types"
+	gammtypes "github.com/osmosis-labs/osmosis/v17/x/gamm/types"
 	gammmigration "github.com/osmosis-labs/osmosis/v17/x/gamm/types/migration"
+	poolManagerTypes "github.com/osmosis-labs/osmosis/v17/x/poolmanager/types"
 	superfluidtypes "github.com/osmosis-labs/osmosis/v17/x/superfluid/types"
 
 	"github.com/osmosis-labs/osmosis/v17/app/keepers"
 	"github.com/osmosis-labs/osmosis/v17/app/upgrades"
 	"github.com/osmosis-labs/osmosis/v17/x/protorev/types"
 )
 
+const (
+	mainnetChainID = "osmosis-1"
+	e2eChainA      = "osmo-test-a"
+	e2eChainB      = "osmo-test-b"
+)
+
+var notEnoughLiquidityForSwapErr = errorsmod.Wrapf(gammtypes.ErrInvalidMathApprox, "token amount must be positive")
+
 func CreateUpgradeHandler(
 	mm *module.Manager,
 	configurator module.Configurator,
 	bpm upgrades.BaseAppParamManager,
 	keepers *keepers.AppKeepers,
 ) upgradetypes.UpgradeHandler {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		poolLinks := []gammmigration.BalancerToConcentratedPoolLink{}
```

Comment:
Don't we need to get this first?

Reply:
No, a pointer to that is passed in to mainnetUpgradeHandler or testnetUpgradeHandler depending on the chain id, which is then filled by its respective logic branch

Reply:
This is because the logic between what we want to migrate differs. Testnet pulls any gamm pool and creates a CL pool for it with a link. Mainnet pulls from the provided list in the constants file. This was so that any testnet can utilize this logic branch (if we specified pools for the testnet logic branch, every testnet we spin up will have to have those pools pre-existing)